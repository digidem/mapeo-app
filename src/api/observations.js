// @flow
import fetch from 'fetch';
import { jsonRequest } from './base';

export interface CreateRequest {
  lat: number;
  lon: number;
}

export type UpdateRequest = {
  id: string
};

class Observation {
  static list = () =>
    jsonRequest({
      method: 'GET',
      route: '/observations'
    });

  static create = (observation: CreateRequest) =>
    jsonRequest({
      method: 'POST',
      route: '/observations',
      body: {
        device_id: '1',
        ...observation
      }
    });

  static update = (observation: UpdateRequest) =>
    jsonRequest({
      method: 'PUT',
      route: `/observations/${observation.id}`,
      body: {
        ...observation
      }
    });
}

export default Observation;
