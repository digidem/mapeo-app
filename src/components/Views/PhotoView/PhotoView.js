// @flow
import React from 'react';
import {
  Dimensions,
  StyleSheet,
  ImageBackground,
  Text,
  TouchableOpacity,
  View
} from 'react-native';
import { NavigationActions, withNavigation } from 'react-navigation';
import Icon from 'react-native-vector-icons/Feather';
import CloseIcon from 'react-native-vector-icons/MaterialIcons';
import type { UpdateRequest } from '@api/observations';
import { CHARCOAL, MAGENTA, MANGO, WHITE } from '../../../lib/styles';
import type { Observation } from '../../../types/observation';
import Gradient from '../../../images/gradient-overlay.png';

export type Props = {
  navigation: NavigationActions
};

export type StateProps = {
  selectedObservation?: Observation
};

export type DispatchProps = {
  updateObservation: (o: UpdateRequest) => void
};

const styles = StyleSheet.create({
  arrowButton: {
    backgroundColor: MANGO,
    height: 40,
    width: 60,
    borderRadius: 25,
    marginRight: 15,
    position: 'absolute',
    bottom: 50
  },
  arrowIcon: {
    alignSelf: 'center',
    justifyContent: 'center',
    marginTop: 2
  },
  backButton: {
    flex: 1,
    backgroundColor: CHARCOAL,
    justifyContent: 'center',
    height: 65
  },
  buttonText: {
    fontSize: 20,
    fontWeight: '700',
    textAlign: 'center',
    color: WHITE
  },
  deleteButton: {
    flex: 1,
    backgroundColor: MAGENTA,
    justifyContent: 'center',
    height: 65
  }
});

class PhotoView extends React.PureComponent<
  Props & StateProps & DispatchProps
> {
  isFromCameraTab() {
    const { navigation } = this.props;

    return !!(
      navigation.state &&
      navigation.state.params &&
      navigation.state.params.fromCameraTab
    );
  }

  isFromDetailView() {
    const { navigation } = this.props;

    return !!(
      navigation.state &&
      navigation.state.params &&
      navigation.state.params.fromDetailView
    );
  }

  handleDeletePhoto = () => {
    const { navigation, updateObservation, selectedObservation } = this.props;

    if (selectedObservation) {
      updateObservation({
        id: selectedObservation.id,
        media: selectedObservation.media.filter(
          photo => navigation.state.params.photoSource !== photo.source
        )
      });
      navigation.goBack();
    }
  };

  render() {
    const { navigation, selectedObservation } = this.props;
    const fromCameraTab = this.isFromCameraTab();
    const fromDetailView = this.isFromDetailView();
    const imageHeight = fromCameraTab
      ? Dimensions.get('window').height
      : Dimensions.get('window').height - 70;
    const hasPhoto =
      selectedObservation &&
      selectedObservation.media &&
      selectedObservation.media.length >= 1;

    return (
      <View style={{ flex: 1 }}>
        <View
          style={{
            flex: 1,
            alignItems: 'center'
          }}
        >
          {hasPhoto &&
            selectedObservation && (
              <ImageBackground
                style={{
                  width: Dimensions.get('window').width,
                  height: fromDetailView
                    ? Dimensions.get('window').height
                    : imageHeight,
                  alignItems: 'center'
                }}
                resizeMode="cover"
                source={
                  navigation.state.params.photoType === 'LocalPhoto'
                    ? navigation.state.params.photoSource
                    : { uri: navigation.state.params.photoSource }
                }
              >
                {fromCameraTab && (
                  <TouchableOpacity
                    style={styles.arrowButton}
                    onPress={() => navigation.navigate('Position')}
                  >
                    <Icon
                      color="white"
                      name="arrow-right"
                      size={35}
                      style={styles.arrowIcon}
                    />
                  </TouchableOpacity>
                )}
                {fromDetailView && selectedObservation.mock ? (
                  <View
                    style={{
                      position: 'absolute',
                      left: 0,
                      backgroundColor: 'transparent'
                    }}
                  >
                    <ImageBackground
                      source={Gradient}
                      style={{
                        width: Dimensions.get('window').width,
                        height: 100
                      }}
                    >
                      <TouchableOpacity
                        style={{ position: 'absolute', right: 15, top: 15 }}
                        onPress={() => navigation.goBack()}
                      >
                        <CloseIcon
                          color="white"
                          name="close"
                          size={30}
                          style={styles.arrowIcon}
                        />
                      </TouchableOpacity>
                    </ImageBackground>
                  </View>
                ) : (
                  <TouchableOpacity
                    style={{ position: 'absolute', right: 15, top: 15 }}
                    onPress={() => navigation.goBack()}
                  >
                    <CloseIcon
                      color="white"
                      name="close"
                      size={30}
                      style={styles.arrowIcon}
                    />
                  </TouchableOpacity>
                )}
              </ImageBackground>
            )}
        </View>
        {!fromCameraTab &&
          !fromDetailView && (
            <View style={{ flexDirection: 'row' }}>
              <TouchableOpacity
                style={styles.backButton}
                onPress={() => navigation.goBack()}
              >
                <Text style={styles.buttonText}>Back</Text>
              </TouchableOpacity>
              <TouchableOpacity
                style={styles.deleteButton}
                onPress={this.handleDeletePhoto}
              >
                <Text style={styles.buttonText}>Delete</Text>
              </TouchableOpacity>
            </View>
          )}
      </View>
    );
  }
}

export default withNavigation(PhotoView);
