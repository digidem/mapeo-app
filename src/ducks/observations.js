// @flow

import update from 'immutability-helper';
import {keyBy} from 'lodash';
import {create} from '../lib/redux';
import {createSelector} from '../lib/selector';
import type {StoreState} from '../types/redux';
import type {Observation} from '../types/observation';

export const {
  type: OBSERVATION_LIST,
  action: observationList,
  reducer: observationListReducer
} = create('OBSERVATION_LIST', {
  success: (state, action) => {
    const observations = keyBy(action.payload, 'id');

    if (state.observations) {
      Object.keys(state.observations).forEach(id => observations[id] = {
        ...observations[id],
        ...state.observations[id],
      });
    }

    const newState = update(state, {observations: {$set: observations}});

    return newState;
  }
});

export const {
  type: OBSERVATION_CREATE,
  action: observationCreate,
  reducer: observationCreateReducer
} = create('OBSERVATION_CREATE', {
  start: (state, action) => {
    const observation = action.meta;

    if (!observation.lat && !observation.lon && state.gps.data) {
      observation.lat = state.gps.data.latitude;
      observation.lon = state.gps.data.longitude;
    }

    const newState = update(state, {selectedObservation: {$set: observation}});

    return newState;
  },
  success: (state, action) => {
    const newState = update(state, {selectObservation: {$set: action.payload}});

    return newState;
  }
});

export const {
  type: OBSERVATION_UPDATE,
  action: observationUpdate,
  reducer: observationUpdateReducer
} = create('OBSERVATION_UPDATE', {
  start: (state, action) => {
    let newState;
    if (state.selectedObservation &&
        action.meta.id === state.selectedObservation.id) {
      newState = update(state, {
        selectedObservation: {
          $set: {...state.selectedObservation, ...action.meta},
        }
      });
    } else {
      newState = update(state, {
        observations: {
          [action.meta.id]: {
            $set: {...state.observations[action.meta.id], ...action.meta}
          }
        }
      });
    }

    return newState;
  }
});

export const {
  type: OBSERVATION_SELECT,
  action: observationSelect,
  reducer: observationSelectReducer
} = create('OBSERVATION_SELECT', {
  start: (state, action) => {
    if (action.meta === undefined && state.selectedObservation) {
      return update(state, {
        selectedObservation: {$set: action.meta},
        observations: {
          [state.selectedObservation.id]: {
            $set: {
              ...(state.observations[state.selectedObservation.id] || {}),
              ...state.selectedObservation,
            }
          }
        }
      });
    } else {
      return update(state, {selectedObservation: {$set: action.meta}});
    }
  }
});

export const selectObservation = createSelector(
    [(state: StoreState, id: string): Observation =>
         state.app.observations[id]],
    (observation: Observation): Observation => observation,
    (observation: Observation, id: string): string => id);

export default [
  observationListReducer, observationCreateReducer, observationUpdateReducer,
  observationSelectReducer
];
