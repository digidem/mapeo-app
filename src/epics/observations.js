// @flow
import React from 'react';
import { Image } from 'react-native';
import type { ActionsObservable } from 'redux-observable';
import {
  OBSERVATION_LIST,
  observationList,
  OBSERVATION_CREATE,
  observationCreate,
  OBSERVATION_UPDATE,
  observationUpdate
} from '../ducks/observations';
import { Action } from '../types/redux';
import type { Observation as ObservationType } from '../types/observation';
import type { StoreState } from '../types/redux';
import Observation from '../api/observations';
import type { CreateRequest, UpdateRequest } from '../api/observations';

export const observationListEpic = (
  action$: ActionsObservable<Action<string, ObservationType[]>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_LIST)
    .filter(action => action.status === 'Start')
    .flatMap(() =>
      Observation.list().map(observations => {
        console.log('RN - Observations List', observations);
        return observationList('', observations);
      })
    );

export const observationCreateEpic = (
  action$: ActionsObservable<Action<CreateRequest, ObservationType>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_CREATE)
    .filter(action => action.status === 'Start')
    .flatMap(action =>
      Observation.create(action.meta).map(observation => {
        console.log('RN - Observation Created');
        return observationCreate(action.meta, observation);
      })
    );

// export const observationUpdateEpic = (
//   action$: ActionsObservable<Action<UpdateRequest, ObservationType>>,
//   store: StoreState
// ) =>
//   action$
//     .ofType(OBSERVATION_UPDATE)
//     .filter(action => action.status === 'Start')
//     .flatMap(action =>
//       Observation.update(action.meta).map(observation => {
//         console.log('RN - Observation Updated - ', action.meta, observation);
//         return observationUpdate(action.meta, observation);
//       })
//     );

export default [
  observationListEpic,
  observationCreateEpic
  // observationUpdateEpic
];
