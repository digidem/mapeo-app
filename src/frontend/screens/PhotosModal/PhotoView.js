// @flow
import React, { useState } from "react";
import { View, StyleSheet, Image, ActivityIndicator } from "react-native";

import { AlertIcon } from "../../sharedComponents/icons";

type Props = {
  uri?: string,
  variant: "photo" | "loading" | "error"
};

const PhotoView = ({ onClose, uri, variant = "photo" }: Props) => {
  const [error, setError] = useState();
  return (
    <View style={styles.container}>
      {variant === "loading" ? (
        <ActivityIndicator />
      ) : variant === "error" || error || !uri ? (
        <AlertIcon />
      ) : (
        <Image
          onError={({ nativeEvent: { error } }) => setError(error)}
          source={{ uri }}
          style={styles.image}
          resizeMethod="scale"
          resizeMode="contain"
        />
      )}
    </View>
  );
};

export default React.memo<Props>(PhotoView);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center"
  },
  image: {
    flex: 1,
    width: "100%",
    height: "100%",
    backgroundColor: "black"
  }
});
