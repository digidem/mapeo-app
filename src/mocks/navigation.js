import type { NavigationScreenProp } from 'react-navigation';

export const createNavigationScreenProp = () => {
  return {
    state: {},
    dispatch: jest.fn(),
    addListener: jest.fn(),
    getParam: jest.fn(),
    isFocused: jest.fn(),
    goBack: jest.fn(),
    navigate: jest.fn(),
    setParams: jest.fn()
  };
};
