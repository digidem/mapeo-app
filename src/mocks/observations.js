// @flow

export const createObservation = (observation?: any) => ({
  type: 'Observation',
  id: 1,
  lat: 50.5,
  lon: 50.5,
  link: 'linkID',
  created: new Date('2018-02-25T20:23:00.000Z'),
  name: 'Oil Spill',
  notes:
    'This is pretty bad. It smells pretty noxious, and the entire pond is affected.',
  observedBy: 'You',
  media: [],
  icon: require('../images/categories/category_18.png'),
  categoryId: 19,

  ...observation
});
