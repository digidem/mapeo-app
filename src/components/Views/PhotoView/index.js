// @flow
import { connect } from 'react-redux';
import type { Dispatch } from 'redux';
import { observationUpdate } from '../../../ducks/observations';

import type { StoreState } from '../../../types/redux';
import PhotoView from './PhotoView';
import type { StateProps, DispatchProps, Props } from './PhotoView';

function mapStateToProps(state: StoreState, ownProps: Props): StateProps {
  const id = ownProps.navigation.getParam('photoId', '');

  return {
    selectedObservation: state.app.selectedObservation,
    attachment: id ? state.app.attachments[id] : undefined
  };
}

function mapDispatchToProps(dispatch: Dispatch<any>): DispatchProps {
  return {
    updateObservation: observation => dispatch(observationUpdate(observation))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(PhotoView);
