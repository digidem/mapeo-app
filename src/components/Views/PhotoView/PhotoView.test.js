// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import PhotoView from './PhotoView';
import { createObservation } from '../../../mocks/observations';
import { createNavigationScreenProp } from '../../../mocks/navigation';
import { resourcePending } from '../../../lib/resource';

describe('PhotoView tests', () => {
  test('snapshot', () => {
    const observation = createObservation();
    const updateObservation = jest.fn();

    const tree = renderer.create(
      <PhotoView
        navigation={createNavigationScreenProp()}
        selectedObservation={observation}
        updateObservation={updateObservation}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
