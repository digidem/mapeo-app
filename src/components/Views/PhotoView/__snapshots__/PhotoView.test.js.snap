// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoView tests snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#1F1F1F",
          "flex": 1,
          "height": 65,
          "justifyContent": "center",
          "opacity": 1,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 20,
            "fontWeight": "700",
            "textAlign": "center",
          }
        }
      >
        Back
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#e2003f",
          "flex": 1,
          "height": 65,
          "justifyContent": "center",
          "opacity": 1,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 20,
            "fontWeight": "700",
            "textAlign": "center",
          }
        }
      >
        Delete
      </Text>
    </View>
  </View>
</View>
`;
